### PR TITLE
Logic Gates

### PR DESCRIPTION
**Pull request by FalseIncarnate**
Adds Logic Gates
- Allows for the performing of logical operations based on signals between the machines, potentially useful for future away mission mapping as it allows us to better design puzzles with buttons and such.

Adds the ability to build and remove light switches!
- 1 metal sheet per switch, just like mass driver buttons.
- They automatically connect to the area they are in when stuck to the wall
- Use a wrench to disconnect an installed light switch from the wall

image

Logic Gates added (from left to right, top row to bottom row):
- NOT
- XOR
- XNOR
- CONVERT
 - Allows for the conversion between signaler and logical signals
- STATUS
 - Emits a colored light related to the logical state of its input
- NOR
- AND
- OR
- NAND

Adds the logic gate circuit board to the AUTOLATHE, so science can't maintain an absurd monopoly on yet another board engineers might want for their forts.

Allows light switches and mass driver buttons to be connected to logic gate inputs via multitool menus.
- Also allows light switches to be disconnected from their area's lights, allowing for logic use without shutting off the lights, or for other reasons

Fixes a potential exploit where mass driver button framess could be deconstructed for more metal than it takes to make them

Cleaned up a bit of relative pathing and also moved some code into more sensible locations

Sprites edited by myself, using some existing frames I found in the files that looked neat.

MAPPER NOTES:
Logic gates can have a "tamperproof" var set on them, to prevent players from directly interacting with the machines (and changing their signals / deconstructing them) and to also render the machine immune to a LOT of potential sabotage sources (like singularities, acid, EMPs, and so forth)

:cl:
rscadd: Adds logic gates, for doing illogical things in a logical manner.
rscadd: Adds buildable light switches, for all your light toggling needs.
bugfix: Fixes a resource duplication bug with mass driver buttons.
/:cl:

### CODE REVIEW

**Comment by FalseIncarnate** [comment]
Weird, those comments from before the PR was even posted carried over.

Anyways, there is one small issue I need help with: attempting to adjust the settings of an attached signaler via the CONVERT Gate doesn't want to display the signaler window. Figured I'll get some help fixing that before this reaches the 24-hour mark, so it seemed fine to post it now.

**Comment by Regen1** [comment]
Why is this a special snowflake board that for some reason doesn't require acid to make? 

There is no monopoly you just have to ask nice

Fuck social interaction amirite?

**Comment by FlattestGuitar** [comment]
Finally the years of redstone experience become useful. 

@Regen1, It's an autolathe board. None of those require acid.

**Comment by Necaladun** [comment]
I like it being accessible outside of science for whatever uses it has.

Minecraft in SS13 when?

**Comment by NTSAM** [comment]
Oh gosh another game where I can get wrapped up by "and or starts ifs" again

File: code/modules/assembly/signaler.dm [hunk 1]
```diff
@@ -18,149 +18,151 @@
...
-		return 1
-
-	update_icon()
-		if(holder)
-			holder.update_icon()
-		return
+/obj/item/device/assembly/signaler/New()
+	..()
+	if(!radio_controller)
+		spawn(40)
```
Comment (tigercat2000) [inline code/modules/assembly/signaler.dm]: could you pretty please replace this with
```
/obj/item/device/assembly/signaler/New()
    . = ..()
    if(radio_controller)
        set_frequency(frequency)

/obj/item/device/assembly/signaler/initialize()
    if(radio_controller)
        set_frequency(frequency)
```

File: code/modules/logic/logic_base.dm [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+
+/obj/machinery/logic_gate/Destroy()
+	if(radio_controller)
+		radio_controller.remove_object(src,frequency)
+	return ..()
+
+/obj/machinery/logic_gate/process()
+	handle_logic()
+	handle_output()		//All output will send for at least one cycle, and will attempt to send every cycle. Hopefully this won't be too taxing.
+	return
```
Comment (tigercat2000) [inline code/modules/logic/logic_base.dm]: Trailing Return.

File: code/modules/logic/converter.dm [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+/obj/machinery/logic_gate/convert/multitool_topic(var/mob/user,var/list/href_list,var/obj/O)
+	..()
+	if("toggle_logic" in href_list)
+		logic_output = !logic_output
+		if(attached_signaler)		//If we have a signaler attached, make sure we update it to send/receive when we change the logic signal desgination
+			if(logic_output)
+				attached_signaler.receiving = 1
+			else
+				attached_signaler.receiving = 0
+	if("adjust_signaler" in href_list && attached_signaler)		//Make sure that we have a signaler attached to handle this one, otherwise ignore this command
```
Comment (Tastyfish) [inline code/modules/logic/converter.dm]: I think this needs to be `if(("adjust_signaler" in href_list) && attached_signaler)`

**Comment by TheDZD** [comment]
Interesting system and concept, hopefully it doesn't end up as under-utilized as a lot of the atmos automation stuff.

**Comment by FalseIncarnate** [comment]
Thanks @Tastyfish for helping with the issue of the signaler window not appearing. It does show up now, though under the multitool menu. That's not really a huge issue, since you can just move the windows.

**Comment by Iamgoofball** [comment]
:-1: for using the radio system

jesus christ that shit is OLD

it needs to stop

**Comment by Iamgoofball** [comment]
:+1: to the rest tho this is good

**Comment by FalseIncarnate** [comment]
One little thing to point out: the circuit boards work like the vendor ones, so you use a screwdriver on the board to change the type of gate it produces. This means there is only one type of board in the auto lathe, which prints out as a random gate board, and you can then adjust it to whatever one you need pretty easily.

**Comment by Tastyfish** [comment]
That sounds highly preferable to having 50 different boards, yeah.

**Comment by ProperPants** [comment]
>the circuit boards work like the vendor ones, so you use a screwdriver on the board to change the type of gate it produces

Oh god... Does it randomize like the vendor boards do too? Because, to the best of my knowledge, it doesn't cycle through vendors with every click, but just randomly selects one, which is fucking agonizing.

**Comment by FalseIncarnate** [comment]
Nope, select from a list. At some point, I'll go and rewrite the vendor board to behave the same

**Comment by TheDZD** [comment]
Code looks alright to me. @Fox-McCloud

**Pull request merged**